Rework captured environment and clear local variables early

mistasse opened the pull request:
This pull request aims at reworking the frame mechanism.

First, the captured environment is stored within a MaterializedFrame associated to the procedure. In this mode, there is thus no more hierarchy of frames. However, the captured environment places itself at the "parent frame" position in the arguments. It allows to reuse the old mechanism with depth of at most 1.
As it opened opportunities to free local variables as soon as they are not used anymore (rather than when leaving the local node for instance) this has also been implemented and is believed to work in most cases, while it may forget to clear one or the other within try-catch blocks. (also in case of a NoElse branch)

mistasse committed:
Rework environment handling

mistasse committed:
Reset unused frame slots

mistasse committed:
Handle variable clearing for AndThen and OrElse

eregon commented on bootcompiler/src/main/scala/org/mozartoz/bootcompiler/transform/VariableClearing.scala:
```diff
@@ -0,0 +1,312 @@
+        val sym = value.asInstanceOf[Variable].symbol
+        if (this.clearsMap.get(sym) == Some(null))
+          this.clearsMap.put(sym, Seq(ClearSite(expr, true)))
+      }
```
Could this be removed and only `merge` used?

eregon replied:
=> the Translator should create its own variables to avoid leaking details here.

eregon commented on bootcompiler/src/main/scala/org/mozartoz/bootcompiler/transform/VariableClearing.scala:
```diff
@@ -0,0 +1,312 @@
+            repl.remove(idx)
+          }
+        }
+      }
```
This should not be needed if `(Variable)` casts in the Translator are replaced by a helper which can see through ClearVarsExpression.

mistasse committed:
Procedure Declaration correctly clears variables it captures

mistasse committed:
Fix ForNode for it to work with variable clearing

eregon commented on bootcompiler/src/main/scala/org/mozartoz/bootcompiler/ast/Statements.scala:
```diff
@@ -116,11 +117,13 @@ case class NoElseStatement()(val pos: Pos) extends Statement with NoElseCommon {
+    val variable = procExpr.args(0)
     val decl = "for " + variable.syntax(indent) + " in " + from.syntax(indent) + ".." + to.syntax(indent) + " do"
-    decl + "\n" + proc.body.syntax(indent + "   ") + "\n" + indent + "end"
+    decl + "\n" + procExpr.body.syntax(indent + "   ") + "\n" + indent + "end"
```
Ah so only ForStatement#syntax depends on it being a ProcExpression :smiley: 
Since this is the only usage, I would do the unwrapping directly in the syntax method:
```scala
val procExpr = WrapperUtil.getInnerExpr(proc).asInstanceOf[ProcExpression]
```

eregon commented on vm/src/org/mozartoz/truffle/nodes/control/ForNode.java:
```diff
@@ -33,6 +33,7 @@ public ForNode(OzNode fromNode, OzNode toNode, OzNode procNode) {
 	public Object execute(VirtualFrame frame) {
 		long from = (long) fromDerefNode.executeDeref(fromNode.execute(frame));
 		long to = (long) toDerefNode.executeDeref(toNode.execute(frame));
+		OzProc proc = (OzProc) procNode.execute(frame);
```
Indeed, I was wondering why I executed it inside the loop (doesn't matter once compiled but it probably does in the interpreter).
This relies on the proc being always the same but of course that's correct for a literal `for` statement.

eregon commented on bootcompiler/src/main/scala/org/mozartoz/bootcompiler/transform/OnStackMarking.scala:
```diff
@@ -30,7 +30,7 @@ sealed class VarState(val onStack: Boolean, val locked: Boolean) {
 }
 object Declared      extends VarState(true, false)
-object Captured      extends VarState(false, false)
+object Captured      extends VarState(false, true) // TODO false false if frame hierarchy, false true if extraction
```
You could read the option in Scala as well to make this correct based on the option value.
It's not ideal but pragmatic for now and that's what I did for self-tail calls in 45edd9a0e3467e9e7bc20409b3cc82097a660f58.

eregon commented on vm/src/org/mozartoz/truffle/nodes/local/ResetSlotsNode.java:
```diff
@@ -0,0 +1,37 @@
+	public Object execute(VirtualFrame frame) {
+		resetSlots(frame, before);
+		Object result = node.execute(frame);
+		resetSlots(frame, after);
```
Should this be a try/finally so that slots are cleared even if `node.execute()` throws an exception ?

eregon committed:
Use createVirtualFrame().materialize() as it is more optimized than createMaterializedFrame()

* createVirtualFrame() is intrinsified by Graal when the descriptor is constant,
  which computes descriptor.getSize() ahead of time and make it constant.
  createMaterializedFrame() ends up in ArrayList#getSize() which is not constant.
  This size is used to allocate primitive and object arrays, as well as tags.

eregon committed:
Use WeakReference when caching variable reads so they do not prevent GC

* WeakReference.get() is folded by Graal in compiled code.

eregon committed:
Implement serialization for the new nodes

eregon committed:
Let Main register Base in the Loader

eregon committed:
Fix CachingReadFrameSlotNode

eregon committed:
Yield to a Coroutine when creating it

* Works around a bug in coroutines which happens when
  terminating a coroutine and switching to a first-time executed coroutine.

eregon commented:
@mistasse Could you take a look at the comments here? I think it's mostly ready except for a few small changes.

eregon commented:
I merged this PR and will apply a few fixes on top, thank you!
